clang/test/CXX/temp/temp.res/p4.cpp

```diff
@@ -185,3 +185,27 @@ template<typename T> struct S {
   friend void X::f(T::type);
 };
 }
+
+namespace GH113324 {
+template <typename = int> struct ct {
+  friend void f1(ct, int = 0);               // expected-error {{friend declaration specifying a default argument must be a definition}}
+  friend void f2(ct a, ct = decltype(a){ }); // expected-error {{friend declaration specifying a default argument must be a definition}}
+};
+
+template<typename T>
+class C {
+public:
+  friend void foo(T a = 1); // expected-error {{friend declaration specifying a default argument must be a definition}}
+};
+
+template<typename T>
+void foo(T a) { } // expected-note {{candidate function template not viable: requires single argument 'a', but no arguments were provided}}
+
+void test() {
+  f1(ct<>{});
+  f2(ct<>{});
+
+  C<int> c;
+  foo(); // expected-error {{no matching function for call to 'foo'}}
```

Comment:
The test I had in mind was something like:
```C++
template<class T> using alias = int;

template<typename T> struct C {
  friend void foo(C, int a = alias<T&>(1));
};

void test() {
  foo(C<void>());
}
```

With your previous patch, we would be able to diagnose the instantiation failure here (forming reference to void), while the current version of the patch doesn't, since it skips instantiation of the default argument.

Reply:
Should the previous changes be reverted?

Reply:
No, I mean not crashing is definitely better. But the error recovery is worse in some situations.

The right fix would be to somehow find the correct function decl pattern for instantiating the default argument.
This would be the function template declaration where that particular default argument was written.

This might not necessarily need to use `getTemplateInstantiationPattern` for that; you could try some other way of finding it. But that ideal fix might be more involved, and I think the current approach is fine as a quick fix, so feel free to continue with this current patch, or otherwise try the better fix, as you wish.

Reply:
I also meant that, whichever way you choose, please add the above test case. In case of keeping the current approach, add a FIXME noting that we miss diagnosing the instantiation failure.